### PR TITLE
support atomic writes for local deep storage

### PR DESCRIPTION
**Pull request by zhxiaogg**
First completed push wins for concurrent pushes to local deep storage.
I also add a simple UT to test the scenario, hope it'll bring any help for review.

### CODE REVIEW

**Comment by zhxiaogg** [comment]
ref #3493

File: api/src/main/java/io/druid/timeline/partition/NoneShardSpec.java [hunk 1]
```diff
@@ -55,6 +56,7 @@ public boolean isInChunk(long timestamp, InputRow inputRow)
   }
 
   @Override
+  @JsonIgnore
```
Comment (drcrallen) [inline api/src/main/java/io/druid/timeline/partition/NoneShardSpec.java]: Add unit tests please

File: server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java [hunk 2]
```diff
@@ -87,19 +92,37 @@ public DataSegment push(File dataSegmentFile, DataSegment segment) throws IOExce
...
+    );
+
+    // moving the temporary directory to the final destination, once success the potentially concurrent push operations
+    // will be failed and will read the descriptor.json created by current push operation directly
+    try {
+      java.nio.file.Files.move(tmpOutDir.toPath(), outDir.toPath());
+    }
+    catch (FileAlreadyExistsException e) {
+      log.warn("Push destination directory[%s] exists, ignore this message if replication is configured.");
+      dataSegment = jsonMapper.readValue(new File(outDir, "descriptor.json"), DataSegment.class);
```
Comment (drcrallen) [inline server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java]: this can throw IOException if I recall correctly, and should be handled properly so that deleteDirectory is cleaned up regardless.

File: server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java [hunk 3]
```diff
@@ -87,19 +92,37 @@ public DataSegment push(File dataSegmentFile, DataSegment segment) throws IOExce
       );
     }
 
+    File tmpOutDir = new File(baseStorageDir, storageDir + "." + UUID.randomUUID().toString());
```
Comment (drcrallen) [inline server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java]: I think a try-with-resources pattern could work here to ensure the temporary directory is cleaned up. Make an `AutoCloseable` (or `Closeable`) that deletes the temporary directory, and wrap the scope of the temporary directory in a `try(AutoCloseable foo = directoryDeletingCloseable)` or similar.

File: server/src/test/java/io/druid/segment/loading/LocalDataSegmentPusherTest.java [hunk 4]
```diff
@@ -48,23 +51,23 @@
...
+    -1
   );
 
   @Before
   public void setUp() throws IOException
   {
     config = new LocalDataSegmentPusherConfig();
     config.storageDirectory = temporaryFolder.newFolder();
-    localDataSegmentPusher = new LocalDataSegmentPusher(config, new ObjectMapper());
+    localDataSegmentPusher = new LocalDataSegmentPusher(config, new JodaMapper());
```
Comment (drcrallen) [inline server/src/test/java/io/druid/segment/loading/LocalDataSegmentPusherTest.java]: Why this change?

File: server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java [hunk 5]
```diff
@@ -69,7 +72,9 @@ public String getPathForHadoop(String dataSource)
   @Override
   public DataSegment push(File dataSegmentFile, DataSegment segment) throws IOException
   {
-    File outDir = new File(config.getStorageDirectory(), DataSegmentPusherUtil.getStorageDir(segment));
+    final String storageDir = DataSegmentPusherUtil.getStorageDir(segment);
+    final File baseStorageDir = config.getStorageDirectory();
+    File outDir = new File(baseStorageDir, storageDir);
```
Comment (drcrallen) [inline server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java]: also suggest `final`

File: server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java (hunk 3 above)
Comment (drcrallen) [inline server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java]: Also suggest `final`

File: server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java [hunk 6]
```diff
@@ -87,19 +92,37 @@ public DataSegment push(File dataSegmentFile, DataSegment segment) throws IOExce
       );
     }
 
+    File tmpOutDir = new File(baseStorageDir, storageDir + "." + UUID.randomUUID().toString());
+    long size = compressSegment(dataSegmentFile, tmpOutDir);
```
Comment (drcrallen) [inline server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java]: suggest `final`

File: server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java [hunk 7]
```diff
@@ -87,19 +92,37 @@ public DataSegment push(File dataSegmentFile, DataSegment segment) throws IOExce
...
+
+    DataSegment dataSegment = createDescriptorFile(
```
Comment (drcrallen) [inline server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java]: suggest `final`

File: server/src/test/java/io/druid/segment/loading/LocalDataSegmentPusherTest.java (hunk 4 above)
Comment (zhxiaogg) [inline server/src/test/java/io/druid/segment/loading/LocalDataSegmentPusherTest.java]: the default mapper cannot deserialize joda Interval which happens when reading `descriptor.json` file.

Comment (drcrallen) [inline server/src/test/java/io/druid/segment/loading/LocalDataSegmentPusherTest.java]: does `DefaultObjectmapper` work?

File: server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java (hunk 2 above)
Comment (zhxiaogg) [inline server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java]: the `push` method signature throws IOException and I think when deletion failed or anything bad happens before deletion, we should keep the "crime site" for later investigation.  This also applies to the above `try-with-resources` suggestion.
Please feel free tell me if I made mistake on this.

File: server/src/test/java/io/druid/segment/loading/LocalDataSegmentPusherTest.java (hunk 4 above)
Comment (zhxiaogg) [inline server/src/test/java/io/druid/segment/loading/LocalDataSegmentPusherTest.java]: Yes, I've run tests with it. I use `JodaMapper` merely because I think it's fast than `DefaultObjectMapper`.

Comment (drcrallen) [inline server/src/test/java/io/druid/segment/loading/LocalDataSegmentPusherTest.java]: For testing purposes consistency and predictability is more important than speed. Since `DefaultObjectMapper` is the de-facto standard `ObjectMapper` for druid code, we should use that whenever possible.

Comment (zhxiaogg) [inline server/src/test/java/io/druid/segment/loading/LocalDataSegmentPusherTest.java]: I agree.

**Comment by zhxiaogg** [comment]
Moved temporary pushes into a intermediate directory right under base persist dir named `intermediate_pushes`.  
 
Code can be reviewed again.

File: server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java (hunk 2 above)
Comment (gianm) [inline server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java]: I really prefer the temp files being kept in the same directory as configured for deep storage, for two reasons:

- For atomic rename of A -> B to work, A and B must be on the same filesystem. A good way to guarantee that is to have them be in the same directory. The java tmp dir is a particularly problematic choice since it's often on a tmpfs.
- Having them in deep storage actually makes them easier to clean up, I think, since operators just need to search through deep storage for files older than a certain date with names matching a certain pattern -- instead of having to track two different directories.

File: server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java [hunk 8]
```diff
@@ -87,19 +92,49 @@ public DataSegment push(File dataSegmentFile, DataSegment segment) throws IOExce
...
+    }
+    catch (FileAlreadyExistsException e) {
+      log.warn("Push destination directory[%s] exists, ignore this message if replication is configured.", outDir);
+      FileUtils.deleteDirectory(tmpOutDir);
+      return jsonMapper.readValue(new File(outDir, "descriptor.json"), DataSegment.class);
+    }
+    return dataSegment;
+  }
+
+  private void createIfNotExists(File directory) throws IOException
```
Comment (drcrallen) [inline server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java]: suggest `createDirectoryIfNotExists` to make it more explicit in what its creating.

**Comment by drcrallen** [comment]
Minor naming change suggestion.

File: server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java (hunk 2 above)
Comment (zhxiaogg) [inline server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java]: @drcrallen I think @gianm is right. I recall that our team owns one special cluster using NFS and we use a script to remove older segments.  If we upgrade the cluster to a new version with this pr, we'll have to change the script to take care of the new `intermediate_pushes` directory.  
On the other hand, if we proceed with `intermediate_pushes` directory, I think we need some explicit doc.
What do you think?

**Comment by zhxiaogg** [comment]
Code could be review again. I'll have to wait may be for further review on `intermediate_pushes` design to finally squash all the commits together.

**Comment by drcrallen** [comment]
Cool! thanks!  Don't worry about squashing, GitHub does that for us now.

File: server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java (hunk 2 above)
Comment (drcrallen) [inline server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java]: I think that's fine

**Comment by zhxiaogg** [comment]
Then all good. Thanks for quick & kind reviewing.

File: server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java [hunk 9]
```diff
@@ -87,19 +92,49 @@ public DataSegment push(File dataSegmentFile, DataSegment segment) throws IOExce
...
+             .withSize(size)
+             .withBinaryVersion(SegmentUtils.getVersionFromDir(dataSegmentFile)),
+      tmpOutDir
+    );
+
+    // moving the temporary directory to the final destination, once success the potentially concurrent push operations
+    // will be failed and will read the descriptor.json created by current push operation directly
+    createDirectoryIfNotExists(outDir.getParentFile());
+    try {
+      java.nio.file.Files.move(tmpOutDir.toPath(), outDir.toPath());
```
Comment (fjy) [inline server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java]: minor nit, but can we just import the lib so we can call Files.move?

**Comment by fjy** [comment]
👍 after my minor change is addressed

**Comment by fjy** [comment]
@zhxiaogg can you address my minor comment and I will merge this
